File: app/src/main/java/com/njust/helper/course/week/CourseView.java
package com.njust.helper.course.week;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.os.Build;
import android.text.Layout;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.njust.helper.R;
import com.njust.helper.model.Course;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by zwb on 2015/4/7.
 * 课程表View
 */
public class CourseView extends View {
    private List<Course>[][] mData;
    private Map<String, Layout> mLayoutContainer = new HashMap<>();
    private int mWeek;
    private TextPaint mTextPaint = new TextPaint();
    private TextPaint mDarkTextPaint = new TextPaint();
    private Paint mLightPaint = new Paint();
    private Paint mDarkPaint = new Paint();
    private Path path;
    private int mLeftColumnSize;
    private int mHeight;
    private int unitHeight, unitWidth;

    private OnSelectCourseListener mListener;

    private int downX, downY;

    public CourseView(Context context) {
        super(context);
        init(null, 0);
    }

    public CourseView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public CourseView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs, defStyleAttr);
    }

    private void init(AttributeSet attrs, int defStyleAttr) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.CourseView, defStyleAttr, 0);
        mHeight = typedArray.getInt(R.styleable.CourseView_courseNum, 5);
        //noinspection unchecked
        mData = new List[7][mHeight];
        mLeftColumnSize = typedArray.getDimensionPixelSize(R.styleable.CourseView_numColumnSize, getResources().getDimensionPixelSize(R.dimen.activity_horizontal_margin));
        typedArray.recycle();

        //noinspection deprecation
        mLightPaint.setColor(0xFFe0e0e0);
        mDarkPaint.setColor(0xFF808080);
        int textSize = getResources().getDimensionPixelSize(R.dimen.course_view_text_size);
        mTextPaint.setTextSize(textSize);
        mTextPaint.setColor(Color.GRAY);
        mTextPaint.setAntiAlias(true);
        mDarkTextPaint.setTextSize(textSize);
        mDarkTextPaint.setColor(Color.BLACK);
        mDarkTextPaint.setFakeBoldText(true);
        mDarkTextPaint.setAntiAlias(true);

        path = new Path();
    }

    public void setCourses(List<Course> courses) {
        if (courses == null) {
            courses = Collections.emptyList();
        }
        for (List[] lists : mData) {
            if (lists != null) {
                for (List list : lists) {
                    if (list != null) {
                        list.clear();
                    }
                }
            }
        }
        for (Course course : courses) {
            List<Course> list = mData[course.getDay()][course.getSec1()];
            if (list == null) {
                list = new ArrayList<>();
                mData[course.getDay()][course.getSec1()] = list;
            }
            list.add(course);
        }
        mLayoutContainer.clear();
        invalidate();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                downX = ((int) event.getX() - mLeftColumnSize) / unitWidth;
                downY = (int) event.getY() / unitHeight;
                return true;
            case MotionEvent.ACTION_UP:
                int x = ((int) event.getX() - mLeftColumnSize) / unitWidth;
                int y = (int) event.getY() / unitHeight;
                if (x == downX && y == downY && mListener != null) {
                    List<Course> list = mData[x][y];
                    if (list != null) {
                        mListener.onSelectCourse(mData[x][y], x, y);
                    }
                }
                return true;
        }
        return super.onTouchEvent(event);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.UNSPECIFIED) {
            setMeasuredDimension(getDefaultSize(getSuggestedMinimumWidth(), widthMeasureSpec),
                    Math.max(getSuggestedMinimumHeight(), MeasureSpec.getSize(heightMeasureSpec)));
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        }
        unitWidth = (getMeasuredWidth() - mLeftColumnSize) / 7;
        unitHeight = getMeasuredHeight() / mHeight;
        int delta2 = unitWidth / 15;
        int side = unitWidth / 4;
        path.moveTo(unitWidth - delta2, unitHeight - delta2 - side);
        path.lineTo(unitWidth - delta2, unitHeight - delta2);
        path.lineTo(unitWidth - delta2 - side, unitHeight - delta2);
        path.close();
    }

    public void setWeek(int week) {
        mWeek = week;
        invalidate();
    }

    @SuppressLint("DrawAllocation")
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

//        int unitWidth = (getWidth() - mLeftColumnSize) / 7;
//        int unitHeight = getHeight() / mHeight;
        float delta = unitWidth / 20;
        for (float i = unitHeight; i < getHeight() + unitHeight; i += unitHeight) {
            for (float j = 0; j < getWidth() - mLeftColumnSize; j += unitWidth) {
                canvas.drawLine(mLeftColumnSize + j - delta, i, mLeftColumnSize + j + delta, i, mTextPaint);
                canvas.drawLine(mLeftColumnSize + j, i - delta, mLeftColumnSize + j, i + delta, mTextPaint);
            }
            canvas.drawText(Integer.toString((int) (i / unitHeight)),
                    delta * 3, i - unitHeight / 2, mTextPaint);
        }

        delta = unitWidth / 30;
        float left = mLeftColumnSize + delta;
        float top = delta;
        canvas.translate(left, top);
        float width = unitWidth - 2 * delta;
        float height = unitHeight - 2 * delta;
        for (int i = 0; i < 7; i++) {
            for (int j = 0; j < mHeight; j++) {
                List<Course> list = mData[i][j];
                Course course = null;
                if (list == null || list.size() == 0) {
                    canvas.translate(0, unitHeight);
                    continue;
                }
                canvas.drawRect(0, 0, width, height, mLightPaint);
                if (list.size() == 1) {
                    course = list.get(0);
                } else {
                    canvas.drawPath(path, mDarkPaint);
                    for (Course t : list) {
                        course = t;
                        if (t.getWeek2().contains(" " + mWeek + " ")) {
                            break;
                        }
                    }
                }
                //noinspection ConstantConditions
                String name = course.getName();
                if (name.length() > 8) {
                    name = name.substring(0, 7) + "...";
                }
                TextPaint paint;
                if (!course.getWeek2().contains(" " + mWeek + " ")) {
                    name = "[非本周]" + name;
                    paint = mTextPaint;
                } else {
                    paint = mDarkTextPaint;
                }
                name = name + "@" + course.getClassroom();
                Layout layout = mLayoutContainer.get(name);
                if (layout == null) {
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                        layout = new StaticLayout(name, paint, (int) (unitWidth - 2 * delta), Layout.Alignment.ALIGN_CENTER, 1f, 0f, false);
                    } else {
                        layout = StaticLayout.Builder.obtain(name, 0, name.length(), paint, (int) (unitWidth - 2 * delta)).build();
                    }
                    mLayoutContainer.put(name, layout);
                }
                layout.draw(canvas);
                canvas.translate(0, unitHeight);
            }
            canvas.translate(unitWidth, -unitHeight * mHeight);
        }
    }

    public void setListener(OnSelectCourseListener listener) {
        mListener = listener;
    }

    public interface OnSelectCourseListener {
        void onSelectCourse(List<Course> courses, int day, int section);
    }
}

File: app/src/main/java/com/njust/helper/course/week/CourseView.kt
package com.njust.helper.course.week

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.os.Build
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.njust.helper.R
import com.njust.helper.model.Course
import java.util.*

typealias OnSelectCourseListener = (courses: List<Course>, day: Int, section: Int) -> Unit

/**
 * Created by zwb on 2015/4/7.
 * 课程表View
 */
class CourseView : View {
    private lateinit var mData: Array<Array<MutableList<Course>>>
    private val mLayoutContainer = HashMap<String, Layout>()
    private var mWeek: Int = 0
    private val mTextPaint = TextPaint()
    private val mDarkTextPaint = TextPaint()
    private val mLightPaint = Paint()
    private val mDarkPaint = Paint()
    private var path: Path = Path()
    private var mLeftColumnSize: Int = 0
    private var mSectionCount: Int = 0
    private var unitHeight: Int = 0
    private var unitWidth: Int = 0

    private var mListener: OnSelectCourseListener? = null

    private var downX: Int = 0
    private var downY: Int = 0

    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int)
            : super(context, attrs, defStyleAttr) {
        init(attrs, defStyleAttr)
    }

    private fun init(attrs: AttributeSet?, defStyleAttr: Int) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.CourseView, defStyleAttr, 0)
        mSectionCount = typedArray.getInt(R.styleable.CourseView_courseNum, 5)

        mData = Array(7) {
            Array(mSectionCount) {
                mutableListOf<Course>()
            }
        }
        mLeftColumnSize = typedArray.getDimensionPixelSize(R.styleable.CourseView_numColumnSize, resources.getDimensionPixelSize(R.dimen.activity_horizontal_margin))
        typedArray.recycle()

        mLightPaint.color = -0x1f1f20
        mDarkPaint.color = -0x7f7f80
        val textSize = resources.getDimensionPixelSize(R.dimen.course_view_text_size)
        mTextPaint.textSize = textSize.toFloat()
        mTextPaint.color = Color.GRAY
        mTextPaint.isAntiAlias = true
        mDarkTextPaint.textSize = textSize.toFloat()
        mDarkTextPaint.color = Color.BLACK
        mDarkTextPaint.isFakeBoldText = true
        mDarkTextPaint.isAntiAlias = true
    }

    fun setCourses(inCourses: List<Course>?) {
        var courses = inCourses
        if (courses == null) {
            courses = emptyList()
        }
        mData.forEach { lists ->
            lists.forEach { list -> list.clear() }
        }
        for (course in courses) {
            var list: MutableList<Course>? = mData[course.day][course.sec1]
            if (list == null) {
                list = ArrayList()
                mData[course.day][course.sec1] = list
            }
            list.add(course)
        }
        mLayoutContainer.clear()
        invalidate()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                downX = (event.x.toInt() - mLeftColumnSize) / unitWidth
                downY = event.y.toInt() / unitHeight
                return true
            }
            MotionEvent.ACTION_UP -> {
                var x = (event.x.toInt() - mLeftColumnSize) / unitWidth
                var y = event.y.toInt() / unitHeight
                if (x == downX && y == downY) {
                    x = Math.min(x, COLUMN_COUNT)
                    y = Math.min(y, mSectionCount)
                    mListener?.invoke(mData[x][y], x, y)
                }
                return true
            }
        }
        return super.onTouchEvent(event)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (View.MeasureSpec.getMode(heightMeasureSpec) == View.MeasureSpec.UNSPECIFIED) {
            setMeasuredDimension(View.getDefaultSize(suggestedMinimumWidth, widthMeasureSpec),
                    Math.max(suggestedMinimumHeight, View.MeasureSpec.getSize(heightMeasureSpec)))
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        }
        unitWidth = (measuredWidth - mLeftColumnSize) / 7
        unitHeight = measuredHeight / mSectionCount
        val delta2 = unitWidth / 15
        val side = unitWidth / 4
        path.moveTo((unitWidth - delta2).toFloat(), (unitHeight - delta2 - side).toFloat())
        path.lineTo((unitWidth - delta2).toFloat(), (unitHeight - delta2).toFloat())
        path.lineTo((unitWidth - delta2 - side).toFloat(), (unitHeight - delta2).toFloat())
        path.close()
    }

    fun setWeek(week: Int) {
        mWeek = week
        invalidate()
    }

    @SuppressLint("DrawAllocation")
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        var delta = (unitWidth / 20).toFloat()
        run {
            var i = unitHeight.toFloat()
            while (i < height + unitHeight) {
                var j = 0f
                while (j < width - mLeftColumnSize) {
                    canvas.drawLine(mLeftColumnSize + j - delta, i, mLeftColumnSize.toFloat() + j + delta, i, mTextPaint)
                    canvas.drawLine(mLeftColumnSize + j, i - delta, mLeftColumnSize + j, i + delta, mTextPaint)
                    j += unitWidth.toFloat()
                }
                canvas.drawText(Integer.toString((i / unitHeight).toInt()),
                        delta * 3, i - unitHeight / 2, mTextPaint)
                i += unitHeight.toFloat()
            }
        }

        delta = (unitWidth / 30).toFloat()
        val left = mLeftColumnSize + delta
        val top = delta
        canvas.translate(left, top)
        val width = unitWidth - 2 * delta
        val height = unitHeight - 2 * delta
        for (i in 0 until COLUMN_COUNT) {
            for (j in 0 until mSectionCount) {
                val list = mData[i][j]
                var course: Course? = null
                if (list.size == 0) {
                    canvas.translate(0f, unitHeight.toFloat())
                    continue
                }
                canvas.drawRect(0f, 0f, width, height, mLightPaint)
                if (list.size == 1) {
                    course = list[0]
                } else {
                    canvas.drawPath(path, mDarkPaint)
                    for (t in list) {
                        course = t
                        if (t.week2.contains(" $mWeek ")) {
                            break
                        }
                    }
                }

                var name = course!!.name
                if (name.length > 8) {
                    name = name.substring(0, 7) + "..."
                }
                val paint: TextPaint
                if (!course.week2.contains(" $mWeek ")) {
                    name = "[非本周]" + name
                    paint = mTextPaint
                } else {
                    paint = mDarkTextPaint
                }
                name = name + "@" + course.classroom
                var layout: Layout? = mLayoutContainer[name]
                if (layout == null) {
                    layout = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                        StaticLayout(name, paint, (unitWidth - 2 * delta).toInt(),
                                Layout.Alignment.ALIGN_CENTER, 1f, 0f,
                                false)
                    } else {
                        StaticLayout.Builder.obtain(name, 0, name.length, paint,
                                (unitWidth - 2 * delta).toInt()).build()
                    }
                    mLayoutContainer.put(name, layout!!)
                }
                layout.draw(canvas)
                canvas.translate(0f, unitHeight.toFloat())
            }
            canvas.translate(unitWidth.toFloat(), (-unitHeight * mSectionCount).toFloat())
        }
    }

    fun setListener(listener: OnSelectCourseListener) {
        mListener = listener
    }

    companion object {
        private const val COLUMN_COUNT = 7
    }
}
